universal-application-tool-0.0.1/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -61,4 +99,59 @@ public ApplicantServiceImpl(
             },
             httpExecutionContext.current());
   }
+
+  /**
+   * Tries to perform an in-place update of {@link ApplicantData}. Returns error messages if it is
+   * impossible to perform the operation.
+   */
+  private ImmutableSet<String> stageUpdates(
+      Applicant applicant,
+      ProgramDefinition programDefinition,
+      long blockId,
+      ImmutableSet<Update> updates) {
+
+    Optional<BlockDefinition> blockDefinitionMaybe = programDefinition.getBlockDefinition(blockId);
+    if (blockDefinitionMaybe.isEmpty()) {
+      return ImmutableSet.of("Block not found.");
+    }
+    BlockDefinition blockDefinition = blockDefinitionMaybe.get();
+
+    ImmutableList<Path> updatePaths =
+        updates.stream().map(Update::path).collect(ImmutableList.toImmutableList());
+    if (!blockDefinition.hasPaths(updatePaths)) {
+      return ImmutableSet.of("Not all paths exist in this Block.");
+    }
+
+    try {
+      stageUpdates(applicant.getApplicantData(), blockDefinition, updates);
+    } catch (UnsupportedScalarTypeException e) {
+      return ImmutableSet.of("Unrecognized scalar type");
+    } catch (PathNotInBlockException e) {
+      return ImmutableSet.of("Update path does not exist.");
```

Comment:
Suggestion (blocking): since these are cases that shouldn't happen in the normal use of the app they can throw, the controller can catch and return a `badRequest`

Reply:
The return is now an `ErrorAnd<ReadOnlyApplicantProgramService, Exception>` for all the different exceptions it may have. The controller is not implemented, but it can check for exceptions and return an appropriate response.